src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Windows.cs

```diff
@@ -61,13 +61,15 @@ public FileSystemEnumerator(string directory, EnumerationOptions options = null)
             _currentPath = _rootDirectory;
 
             int requestedBufferSize = _options.BufferSize;
-            int bufferSize = requestedBufferSize <= 0 ? StandardBufferSize
+            _bufferLength = requestedBufferSize <= 0 ? StandardBufferSize
                 : Math.Max(MinimumBufferSize, requestedBufferSize);
 
             try
             {
-                _buffer = ArrayPool<byte>.Shared.Rent(bufferSize);
-                _pinnedBuffer = GCHandle.Alloc(_buffer, GCHandleType.Pinned);
+                // NtQueryDirectoryFile needs its buffer to be 64bit aligned to work
+                // successfully with FileFullDirectoryInformation on ARM32. AllocHGlobal
+                // will return pointers aligned as such, new byte[] does not.
+                _buffer = Marshal.AllocHGlobal(_bufferLength);
```

Comment:
I'm wondering - is the issue specific to the `ArrayPool<byte>.Shared.Rent` or every `new byte[]` - I'd suspect way more code to be broken if `byte[]` was giving non-aligned pointers when marshalling

Reply:
> every `new byte[]`

Correct.

>  I'd suspect way more code to be broken if byte[] was giving non-aligned pointers when marshalling

Why do you think so? It is pretty unusual to take `byte[]`, pin it and use it as a structure with non-trivial alignment requirements. If you need to pass a struct to unmanaged, you typically allocate the struct directly, not by going via `byte[]`.